lib/vast/Conversion/FromHL/LowerEnums.cpp

```diff
@@ -0,0 +1,177 @@
+
+        llvm::DenseMap< llvm::StringRef, const_decl > id_to_value;
+        llvm::DenseMap< llvm::StringRef, enum_decl > name_to_decl;
+
+        enums_info(vast_module mod) {
+            collect_enum_values(mod);
+        }
+
+        void collect_enum_values(vast_module root) {
+            auto walker = [&](operation op) {
+                if (auto decl = mlir::dyn_cast< const_decl >(op)) {
+                    VAST_ASSERT(!id_to_value.count(decl.getName()));
+                    id_to_value[decl.getName()] = decl;
+                } else if (auto decl = mlir::dyn_cast< enum_decl >(op)) {
+                    name_to_decl[decl.getName()] = decl;
+                }
+            };
+
+            root->walk(walker);
+        }
+    };
+
+    struct enum_type_converter
+        : tc::identity_type_converter
+        , tc::mixins< enum_type_converter >
+    {
+        mcontext_t &mctx;
+        const enums_info &info;
+
+        enum_type_converter(mcontext_t &mctx, const enums_info &info)
+            : mctx(mctx), info(info)
+        {
+            init();
+        }
+
+        auto convert_enum() {
+            return [&](hl::RecordType record) -> maybe_type_t {
+                // If not in the mapping it is not an `enum` (both `union` and `struct` use
+                // `record` as well.
+                auto name = record.getName();
+                auto it = info.name_to_decl.find(name);
+                if (it == info.name_to_decl.end())
+                    return {};
+
+                // Need to be a separate decl to avoid compiler thinking it is a `const`.
+                hl::EnumDeclOp decl = it->second;
+                auto type = decl.getType();
+                VAST_CHECK(type, "enum did not provide `getType()`, {0}", decl);
+                return type;
+            };
+        }
+
+        void init() {
+            addConversion(convert_enum());
+        }
+    };
+
+    struct type_rewriter : pattern_rewriter {
+        type_rewriter(mcontext_t *mctx) : pattern_rewriter(mctx) {}
+    };
+
+    using type_conversion_pattern = tc::generic_type_converting_pattern< enum_type_converter >;
+
+    struct LowerEnumsPass : LowerEnumsBase< LowerEnumsPass > {
+        void runOnOperation() override {
+            auto enums = enums_info(this->getOperation());
+
+            // Investigate if we want these as patterns instead. I would be curious
+            // to see what's the runtime difference of repeating the traversal vs the
+            // bookkeeping the conversion needs to work.
+            // For now I am keeping this separate in case some random edge cases needs to be
+            // added (this is easier to debug),
+            // but eventually we probably should swap these to patterns.
+            replace_constants(enums);
+            replace_var_init(enums);
+            replace_types(enums);
+
+            erase_enums(enums);
+        }
+
+        void replace_constants(const enums_info &enums) {
```

Comment:
This should be cannonicalization pass (i.e. just definition how to emit constant from enum op) rather than pass.